File: bindings/elixir/lib/umicp/transport/http2.ex
defmodule UMICP.Transport.HTTP2 do
  @moduledoc """
  HTTP/2 transport for UMICP using Finch.
  """

  alias UMICP.Envelope

  @doc """
  Send an envelope via HTTP/2 POST.
  """
  @spec send(String.t(), Envelope.t(), keyword()) ::
          {:ok, Envelope.t()} | {:error, term()}
  def send(base_url, envelope, opts \\ []) do
    endpoint = Keyword.get(opts, :endpoint, "/umicp")
    url = base_url <> endpoint

    with {:ok, json} <- Envelope.serialize(envelope),
         {:ok, %Finch.Response{status: status, body: body}} when status in 200..299 <-
           Finch.build(:post, url, [{"content-type", "application/json"}], json)
           |> Finch.request(UMICPFinch),
         {:ok, response_envelope} <- Envelope.deserialize(body) do
      {:ok, response_envelope}
    else
      {:ok, %Finch.Response{status: status}} ->
        {:error, {:http_error, status}}

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Send GET request and receive envelope.
  """
  @spec get(String.t(), String.t()) :: {:ok, Envelope.t()} | {:error, term()}
  def get(base_url, endpoint) do
    url = base_url <> endpoint

    with {:ok, %Finch.Response{status: status, body: body}} when status in 200..299 <-
           Finch.build(:get, url)
           |> Finch.request(UMICPFinch),
         {:ok, envelope} <- Envelope.deserialize(body) do
      {:ok, envelope}
    else
      {:ok, %Finch.Response{status: status}} ->
        {:error, {:http_error, status}}

      {:error, reason} ->
        {:error, reason}
    end
  end
end

File: bindings/swift/Sources/UMICP/Transport/HTTP2.swift
import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// HTTP/2 transport for UMICP with async/await support
@available(macOS 12.0, iOS 15.0, *)
public struct UMICPHTTP2 {
    private let baseURL: URL
    private let session: URLSession
    
    public init(baseURL: URL) {
        self.baseURL = baseURL
        
        let configuration = URLSessionConfiguration.default
        configuration.httpAdditionalHeaders = [
            "Content-Type": "application/json",
            "Accept": "application/json"
        ]
        configuration.httpMaximumConnectionsPerHost = 10
        
        self.session = URLSession(configuration: configuration)
    }
    
    /// Send envelope via HTTP/2 POST
    public func send(envelope: Envelope, endpoint: String = "/umicp") async throws -> Envelope {
        let url = baseURL.appendingPathComponent(endpoint)
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        
        let json = try envelope.serialize()
        request.httpBody = json.data(using: .utf8)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw UMICPError.transportError("Invalid response type")
        }
        
        guard (200...299).contains(httpResponse.statusCode) else {
            throw UMICPError.transportError("HTTP error: \(httpResponse.statusCode)")
        }
        
        guard let responseText = String(data: data, encoding: .utf8) else {
            throw UMICPError.serializationError("Invalid UTF-8 response")
        }
        
        return try Envelope.deserialize(from: responseText)
    }
    
    /// Send GET request and receive envelope
    public func get(endpoint: String) async throws -> Envelope {
        let url = baseURL.appendingPathComponent(endpoint)
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw UMICPError.transportError("Invalid response type")
        }
        
        guard (200...299).contains(httpResponse.statusCode) else {
            throw UMICPError.transportError("HTTP error: \(httpResponse.statusCode)")
        }
        
        guard let responseText = String(data: data, encoding: .utf8) else {
            throw UMICPError.serializationError("Invalid UTF-8 response")
        }
        
        return try Envelope.deserialize(from: responseText)
    }
}

